[Emotion] Logical property, padding, and background color utilities

cchaos opened the pull request:
### These added utilities are some of the most commonly used/re-used properties among EUI and consumer styles.

## Logical properties

These help convert standard direction properties like `left`, `height`, and `margin-bottom` into logical properties like `inline-start`, `inset-block`, and `margin-block-end`.

**Docs lives under Theming/SIze: https://eui.elastic.co/pr_5850/#/theming/sizing#utilities**

**NOTE: ** I did not prefix these helpers with `eui` because they're generic and not specific to EUI in any way.

- `logicals{}`: Map of all the directional properties to the logical counterpart
- ~`logicalCSS()`: Function to return the string version of the `property: value` pair~
- `logicalStyle()`: Function to return the object version of the `property: value` pair
- `logicalTextAlign()`: Text-specific function because it's the value that needs to change not the property


## Padding

These helpers either create the full map of common padding sizes (styles function) including logical side or quickly grab one of the pre-defined padding attributes based on side and size (hook).

**Docs lives under Theming/Size: https://eui.elastic.co/pr_5850/#/theming/sizing#utilities**

- `euiPaddingSize()`: Returns the calculated padding size value only
- `useEuiPaddingSize()`: Hook version of ^^
- `useEuiPaddingStyles()`: Returns the full map of size to padding value for quick styling of components that can have their padding size changed via a prop. It also accepts a `side` that will update the property based on the logical property equivelant.

## Background color

Similar to the padding functions, these create maps specifically around colorizing background colors based off the brand colors (lightly shaded versions). It's similar to how EuiPanel colors it's background but now in a re-usable style fashion.

**Docs lives under Theming/Color: https://eui.elastic.co/pr_5850/#/theming/colors#utilities**

- `euiBackgroundColor()`: just calculates and returns the color
- `useEuiBackgroundColor()`: Hook that returns the string version calculated color
- `useEuiBackgroundColorStyles()`: Returns the full map of token to  `background-color: value` pair for quick styling of components that can have their background color changed via a prop

**Contrast specific docs added to the bottom of the contrast tab.**

<img width="946" alt="Screen Shot 2022-05-02 at 01 13 26 AM" src="https://user-images.githubusercontent.com/549577/166187564-1af34d3c-3199-4f73-873d-a77880e5becd.png">



### Checklist

- [x] Checked in both **light and dark** modes
- ~[ ] Checked in **mobile**~
- [x] Checked in **Chrome**, **Safari**, **Edge**, and **Firefox**
- ~[ ] Props have proper **autodocs** and **[playground toggles](https://github.com/elastic/eui/blob/main/wiki/documentation-guidelines.md#adding-playground-toggles)**~
- [x] Added **[documentation](https://github.com/elastic/eui/blob/main/wiki/documentation-guidelines.md)**
- ~[ ] Checked **[Code Sandbox](https://codesandbox.io/)** works for any docs examples~
- ~[ ] Added or updated **[jest](https://github.com/elastic/eui/blob/main/wiki/testing.md) and [cypress](https://github.com/elastic/eui/blob/main/wiki/cypress-testing.md) tests**~
- ~[ ] Checked for **breaking changes** and labeled appropriately~
- ~[ ] Checked for **accessibility** including keyboard-only and screenreader modes~
- [x] Updated the **[Figma](https://www.figma.com/community/file/964536385682658129)** library counterpart
- [x] A **[changelog](https://github.com/elastic/eui/blob/main/wiki/documentation-guidelines.md#changelog)** entry exists and is marked appropriately

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5850/

cchaos commented on src/components/horizontal_rule/horizontal_rule.styles.ts:
```diff
@@ -7,28 +7,29 @@
   `,
   quarter: css`
-    width: 25%;
-    margin-inline: auto;
+    ${logicalCSS('width', '25%')}
+    ${logicalCSS('margin-horizontal', 'auto')}
   `,
```
I've started updating some properties, but at the end of the day, if we know/remember the logical property as we see below with `margin-block` it's somewhat easier to read. So I 🤷‍♀️ about whether we **need** to use these helpers everywhere or not. The two arguments I see **for** using the util are:

1. Future proof readability for folks not fully knowledgable of the new properties
2. If the CSS spec ever changes, we have one place we need to update (least likely scenario at this point)

cchaos commented on src/global_styling/mixins/_padding.ts:
```diff
@@ -0,0 +1,44 @@
+export const PADDING_SIZES = ['none', 'xs', 's', 'm', 'l', 'xl'] as const;
+export type EuiPaddingSize = typeof PADDING_SIZES[number];
+
+export const euiPaddingStyles = (
```
This is basically the generic equivalent/conversion of `$euiPanelPaddingModifiers`

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5850/

cchaos commented:
I took another page out of @constancecchen's book (i.e. copied her work) and created tests for these utils. This is a great method also to quickly compare the values of every acceptable parameter (and makes it easy for me to update Figma 😉 )

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5850/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5850/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5850/

miukimiu commented on upcoming_changelogs/5850.md:
```diff
@@ -0,0 +1,3 @@
+- Added `logicals{}`, `logicalCSS()`, `logicalStyle()`, and `logicalTextAlign()` CSS property utilities
+- Added `euiPaddingSize()`, `useEuiPaddingSize()`, and `useEuiPaddingStyels()` sizing utilities
```
```suggestion
- Added `euiPaddingSize()`, `useEuiPaddingSize()`, and `useEuiPaddingStyles()` sizing utilities
```

miukimiu reviewed (changes requested):
Thanks, @cchaos. It's looking good but I found a few issues.

In the `theming/colors` contrast tab the tooltips are showing `undefined` for the background colors in **Body background color**, **Brand colors** and in **Shades**:

<img width="634" alt="Screenshot 2022-05-02 at 13 12 43" src="https://user-images.githubusercontent.com/2750668/166232965-d76a9e74-f098-4fe6-9575-876ed90e07b1.png">

In **Background colors** the `useEuiBackgroundColor()` argument should be a `string`. So when we click copy this is what is copied:

```js
// the `accent` should be a string
background-color: ${useEuiBackgroundColor(accent)};
color: ${euiTheme.colors.primaryText};
```
<img width="356" alt="Screenshot 2022-05-02 at 13 24 33" src="https://user-images.githubusercontent.com/2750668/166233476-cba8ca65-1a6b-4bdb-9516-0afa94df9509.png">

cchaos commented:
Good catch on the previous contrast badges 💟 ! Thank you!

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5850/

constancecchen commented on src/global_styling/functions/logicals.ts:
```diff
@@ -0,0 +1,121 @@
+
+import { keysOf } from '../../components/common';
+
+export const logicalSide = {
```
Any chance we can add a link to either https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Logical_Properties or https://css-tricks.com/css-logical-properties-and-values/ for those of us new to CSS logical properties? 🙏

cchaos replied:
I added one in our docs, but I can add in this file too

constancecchen replied:
Yeah I just saw the link in the docs! Up to you and this is def not a blocking request on my part, but I'd personally appreciate a super quick code comment link here as well since I tend to try to look up definitions super quickly in VSCode vs. opening a new tab to our docs, especially for internal helpers

constancecchen commented on src/global_styling/functions/logicals.ts:
```diff
@@ -0,0 +1,121 @@
+ * @param value String to output as the property value
+ * @returns `string` Returns the logical CSS property version for the given `property: value` pair
+ */
+export const logicalCSS = (property: LogicalProperties, value?: any) => {
```
I think we can be more specific here with the `value` and type it as `string`, based on the options I'm seeing on various docs - WDYT? Also, does it really make sense for `value` to be optional? When would we ever want to output CSS that's empty?

```suggestion
export const logicalCSS = (property: LogicalProperties, value: string) => {
```

cchaos replied:
I did have it as `string` once, but then I ran into this conflicting TS issue:
<img width="850" alt="Screen Shot 2022-05-02 at 15 56 27 PM" src="https://user-images.githubusercontent.com/549577/166316516-85a281c6-07ef-455c-901d-2caad883b440.png">

Basically when we set a lot of keys to be `CSSProperties` based, users can also just use simple numbers. So we can't be certain whether `value` will be exactly a string. I felt it can be fairly tolerable and they'd have to understand CSS values. 

Otherwise we'll have to go in and change some of our global types.

---

I also had it optional in case someone setup of an array and mapped through values including a `none: undefined` option.

constancecchen replied:
Ahhh gotcha. No worries then if it's a downstream type headache.

constancecchen commented on src/global_styling/functions/logicals.ts:
```diff
@@ -0,0 +1,121 @@
+ * @param value String to output as the property value
+ * @returns `object` Returns the logical CSS property version for the given `property: value` pair
+ */
+export const logicalStyle = (property: LogicalProperties, value?: any) => {
```
See above comment

```suggestion
export const logicalStyle = (property: LogicalProperties, value: string) => {
```

constancecchen commented on src/global_styling/functions/logicals.ts:
```diff
@@ -0,0 +1,121 @@
+ * @returns `string` Returns the logical CSS property version for the given `property: value` pair
+ */
+export const logicalCSS = (property: LogicalProperties, value?: any) => {
+  return `${logicals[property]}: ${value};`;
+};
+
+/**
+ *
+ * @param property A string that is a valid CSS logical property
+ * @param value String to output as the property value
+ * @returns `object` Returns the logical CSS property version for the given `property: value` pair
+ */
+export const logicalStyle = (property: LogicalProperties, value?: any) => {
```
Maybe I'm missing something, but why bother having two separate string/object versions of the same utility? IIRC Emotion's `css` util handles automatically converting objects inside string literals, so we should only need the object util.

constancecchen replied:
I pushed up a regex fix for the React error but I'm still not sure that fully addresses my question here - I'm saying we only need `logicalStyle` and not `logicalCSS`, as the object should work for both the `style` and `css` props (thanks to Emotion magic). Thoughts?

cchaos replied:
So originally I didn't know Emotion converted objects, so 🙌 .

However, I guess this is going to be a generalized question for (most) of these mixins. We **should** support any method of styling from Emotion strings, to React `style` props, to styled-components. But right now they don't all really play nice. The way we've been handling it by returning a simple string has been really simple and easy. 

It wasn't until I realized that we'd want these specific logical helpers for `style` properties that I ran into that snag with the string format. TBH, it's frustrating React behaves this way for what are completely valid HTML `style` properties.

But if we know we're writing a mixin that could be used for `style` tags, we have to be able to pass back the object version. But as you saw with having to convert from `snake-case` to `camelCase` that it gets messy. Not only that but also makes it harder to just see the output of function because it's now returning an object instead of a string:

```tsx
{Object.entries(logicalStyle('padding-left', '160px'))[0][0]}:{' '}
{Object.entries(logicalStyle('padding-left', '160px'))[0][1]};
```

Unless there's an easier way that I'm not aware of. So instead of asking for another parameter of `string` vs `object` as the return, I opted for two separate functions that could easily be expanded on and kept isolated where having to do that key manipulation only happens **if** you need the object (which is probably going to be the least common usage).

constancecchen replied:
> The way we've been handling it by returning a simple string has been really simple and easy.

Ahh maybe I'm confused then, I was actually going off of the `euiFontSize()` util only returning an object and thinking that was primarily what we wanted for mixins that return euiTheme based content.

> We should support any method of styling from Emotion strings, to React style props, to styled-components.

FWIW [styled-components does support objects](https://styled-components.com/docs/advanced#style-objects) it looks like so now I'm wondering if objects might be the most agnostic approach for all possible platforms.

> Not only that but also makes it harder to just see the output of function because it's now returning an object instead of a string:

Hmm, I'm not following the above objection. I don't think objects are any more difficult to examine than strings. In general we would "see" the final output by inspecting the CSS in devtools, but console logging objects should not really be an issue, or using snapshots in the case of unit tests. In a pinch I would also opt to `JSON.stringify(object)` to inspect it if for some reason the debugger isn't expanding it. Or am I missing what you mean when you say "see the output"?

cchaos replied:
`JSON.stringify`! I forgot about that. Not as clean as just the `property: value` pair output (in the docs) but maybe it's better to be **very** clear what is being passed back.
<img width="994" alt="Screen Shot 2022-05-02 at 19 43 34 PM" src="https://user-images.githubusercontent.com/549577/166343296-b6cd4228-9e4e-4459-87f7-fc91617f998e.png">


In, https://github.com/elastic/eui/pull/5850/commits/ded07abe11966e838164ce18b88bc775918350b3, I've opted for your suggestion of leaving to just `logicalStyle` and removed `logicalCSS`. We'll see if there's other effects as we continue to convert.

constancecchen commented on src/components/loading/loading_chart.styles.ts:
```diff
@@ -36,15 +36,15 @@ export const euiLoadingChartBarStyles = ({ euiTheme }: UseEuiTheme) => ({
   m: css`
     width: ${euiTheme.size.xxs};
-    margin-block-end: ${euiTheme.size.s};
+    ${logicalCSS('margin-bottom', euiTheme.size.s)};
   `,
   l: css`
     width: ${euiTheme.size.xs};
-    margin-block-end: ${euiTheme.size.m};
+    ${logicalCSS('margin-bottom', euiTheme.size.m)};
   `,
   xl: css`
     width: ${euiTheme.size.s};
-    margin-block-end: ${euiTheme.size.base};
+    margin-${logicalSide.bottom}: ${euiTheme.size.base};
```
Is there a specific reason why this is third example property is different from the first 2? 🤔

cchaos replied:
Mainly that I was testing that they work and seeing the difference in formatting, wondering if people had a preference or if the preference just lies with the author of the styles.

constancecchen commented on src/global_styling/mixins/__snapshots__/_padding.test.ts.snap:
```diff
@@ -0,0 +1,139 @@
+      padding-block: 4px;
+    ",
+}
+`;
```
These snapshots were super useful, thanks for including them! ✨

constancecchen commented on src/global_styling/functions/logicals.test.ts:
```diff
@@ -0,0 +1,55 @@
+  });
+});
+
+describe('logicalTextAlign mixin returns an string property', () => {
```
super minor grammar nit
```suggestion
describe('logicalTextAlign mixin returns a string property', () => {
```

constancecchen commented on src/global_styling/functions/logicals.test.ts:
```diff
@@ -0,0 +1,55 @@
+      expect(
+        testCustomHook(() => logicalTextAlign('right')).return
+      ).toMatchSnapshot();
+    });
```
Just for completeness, should we also test `center` in here as well? I'd also be open to exporting the 3 possible values as an array and making this a forEach

cchaos replied:
Ah yes, I forgot I was going to create those types. On it.

constancecchen commented:
I mostly just reviewed the source code changes skimmed the documentation section since Elizabet already did a great job reviewing it - changes look great! I think https://github.com/elastic/eui/pull/5850#discussion_r863107118 is my only major/blocking question, all other comments are minor nits/suggestions.

cchaos commented on src-docs/src/views/theme/sizing/_sizing_js.tsx:
```diff
@@ -122,3 +134,147 @@ export default () => {
+          </p>
+        }
+        example={
+          <p style={logicalStyle('padding-left', '160px')}>
```
@constancecchen Here I mention how I'm still having issues with React parsing the object version of the function. Maybe you could lend some help?

constancecchen replied:
Pushed up a fix! https://github.com/elastic/eui/pull/5850/commits/4a3a034692b48c6fba600d1d430ea0f22e65db7a

cchaos replied:
Thank you!!

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5850/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5850/

cchaos commented:
I think I've addressed all concerns. Thank you both! 💟 

P.S. I'll update the docs' sections with parameter types when #5855 merges to utilize the new ThemeValue prop added there.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5850/

constancecchen commented:
Oh no! I'm not sure what's going wrong, but it looks like Emotion isn't rendering the expected output on https://eui.elastic.co/pr_5850/#/display/avatar 😬 

Maybe it's the semi-colons following the mixin usage? I'm not actually sure they should be included when returning an object

cchaos commented:
Womp womp. It's not just the semicolons. It seems passing back an object expects it to be very precisely applied. After removing them, I still got the following css output:

```css
    overflow-block: hidden;
    font-weight: 500;
    [object Object] [object Object] line-height: calc(16px * 4);
```

It's because of the extra wrapping function that was supposed to just return a string: https://github.com/elastic/eui/blob/ded07abe11966e838164ce18b88bc775918350b3/src/components/avatar/avatar.styles.ts#L20-L22

I can change this particular instance to return a serialized string, but I think we may run into this more often than not since we're trying to always return strings.

That's why I think optimizing for the string value is better than optimizing for the object. Still begs the question though of whether we need to have two mixins per.

constancecchen commented:
D'oh. Totally makes sense. I'd be good with consistently returning strings personally. I do think we should prefer to pick 1 or the other for developer experience. @thompsongl any thoughts here?

constancecchen commented:
Per our sync today, I'd be good with reverting [ded07ab](https://github.com/elastic/eui/pull/5850/commits/ded07abe11966e838164ce18b88bc775918350b3) and shipping this PR with both string and object utils!

cchaos commented:
Sorry about the commit history, but this PR should be updated now based on the discussion. 😅

constancecchen reviewed (approved):
🎉

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5850/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5850/
